angr/analyses/vtable.py

```diff
@@ -0,0 +1,67 @@
+from . import Analysis
+
+
+class Vtable:
+    def __init__(self, vaddr, size, func_addrs=None):
+        self.vaddr = vaddr
+        self.size = size
+        self.func_addrs = func_addrs if func_addrs else []
+
+
+class VtableFinder(Analysis):
+    def __init__(self):
+        self.cfg = self.project.analyses.CFGFast(cross_references=True)
+        self.vtables_list = self.analyze()
+
+    def is_cross_referenced(self, addr):
+        if addr in self.cfg.kb.xrefs.xrefs_by_dst:
+            return True
+        else:
+            return False
+
+    def is_function(self, addr):
+        if addr in self.cfg.kb.functions:
+            return True
+        else:
+            return False
+
+    def analyze(self):
+        # finding candidate starting vtable addresses
+        # "current location is referenced from a code segment andits value is a pointer to a function,
+        # then it is marked as astart of vtable"
+        # taken from - Reconstruction of Class Hierarchies for Decompilation of C++ Programs
+        list_vtables = []
+        for sec in self.project.loader.main_object.sections:

```

Comment:
Can we make this use segments instead of sections? Or even better yet, use segments if they're present and fall back to sections if not?

Reply:
I tried that now and running the analysis on read only segments resulted in a lot of false positive VTables. I think the heuristic works better when we use just these sections `.data.rel.ro`, `.rodata`, `.data.rel.ro.local`

Reply:
Disable this detection if sections do not exist.